src/KubernetesClient/Authentication/GcpTokenProvider.cs

```diff
@@ -0,0 +1,57 @@
+using System;
+using System.Diagnostics;
+using System.Net.Http.Headers;
+using System.Threading;
+using System.Threading.Tasks;
+using k8s.Exceptions;
+using Microsoft.Rest;
+using Newtonsoft.Json.Linq;
+
+namespace k8s.Authentication
+{
+    public class GcpTokenProvider : ITokenProvider
+    {
+        private readonly string _gcloudCli;
+        private string _token;
+        private DateTime _expiry;
+
+        public GcpTokenProvider(string gcloudCli)
+        {
+            _gcloudCli = gcloudCli;
+        }
+
+        public Task<AuthenticationHeaderValue> GetAuthenticationHeaderAsync(CancellationToken cancellationToken)
+        {
+            if (DateTime.UtcNow.AddSeconds(30) > _expiry)
+            {
+                RefreshToken();
+            }
+            return Task.FromResult(new AuthenticationHeaderValue("Bearer", _token));
+        }
+
+        private void RefreshToken()
```

Comment:
Shouldn't this be async and use an async command line runner?

Reply:
Switched to non blocking command line runner